core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java

```diff
@@ -13,20 +13,89 @@
  */
 package org.jdbi.v3.core.statement;
 
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
 import java.sql.PreparedStatement;
 import java.sql.SQLException;
 import java.util.Set;
 
-import org.jdbi.v3.core.argument.NullArgument;
+import org.jdbi.v3.core.argument.Argument;
+import org.jdbi.v3.core.internal.exceptions.Unchecked;
 
 class DefineNamedBindingsStatementCustomizer implements StatementCustomizer {
     @Override
-    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) throws SQLException {
+    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) {
         final Set<String> alreadyDefined = ctx.getAttributes().keySet();
         final Binding binding = ctx.getBinding();
+        final SetNullHandler handler = new SetNullHandler();
         binding.getNames().stream()
             .filter(name -> !alreadyDefined.contains(name))
             .forEach(name -> binding.findForName(name, ctx).ifPresent(
-                    a -> ctx.define(name, a instanceof NullArgument ? false : true)));
+                    a -> handler.define(name, a, ctx)));
+    }
+
+    private static class SetNullHandler implements InvocationHandler {
+        private final PreparedStatement fakeStmt = (PreparedStatement)
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws SQLException {
+            if (method.getName().equals("unwrap")
+                && args.length == 1
+                && method.getParameterTypes()[0].equals(Class.class)) {
+                throw new SQLException("The current implementation of DefineNamedBindings is incompatible with "
+                    + "arguments that rely on java.sql.Wrapper.unwrap(Class<?>)");
+            }
+
+            if (method.getName().startsWith("set")) {
+                setCalled = true;
+                boolean argNull = args.length > 1 && args[1] == null;
+                setNull = argNull || "setNull".equals(method.getName());
+            }
+
+            return defaultValue(method.getReturnType());
+        }
+
+        private Object defaultValue(Class<?> type) {
+            if (type.isPrimitive()) {
```

Comment:
It'd be possible to shorten this significantly: `Array.get(Array.newInstance(type, 1), 0)` but maybe that's too obscure

Reply:
This smells of a Map<Class, Object> to me